cmd/cert-checker/main.go

```diff
@@ -0,0 +1,260 @@
+// Copyright 2015 ISRG.  All rights reserved
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package main
+
+import (
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"runtime"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/cactus/go-statsd-client/statsd"
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/codegangsta/cli"
+	gorp "github.com/letsencrypt/boulder/Godeps/_workspace/src/gopkg.in/gorp.v1"
+	"github.com/letsencrypt/boulder/policy"
+
+	"github.com/letsencrypt/boulder/cmd"
+	"github.com/letsencrypt/boulder/core"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/sa"
+)
+
+const (
+	good = "valid"
+	bad  = "invalid"
+
+	filenameLayout = "20060102"
+
+	checkPeriod = time.Hour * 24 * 90
+
+	batchSize = 100000
+)
+
+type report struct {
+	GoodCerts int64
+	BadCerts  int64
+	Entries   map[string]reportEntry
+}
+
+func (r *report) save(directory string) error {
+	filename := path.Join(directory, fmt.Sprintf(
+		"%s-%s-report.json",
+		time.Now().Add(-checkPeriod).Format(filenameLayout),
+		time.Now().Format(filenameLayout),
+	))
+	content, err := json.Marshal(r)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(filename, content, os.ModePerm)
+}
+
+type reportEntry struct {
+	Valid    bool     `json:"validity"`
+	Problems []string `json:"problem,omitempty"`
+}
+
+type certChecker struct {
+	pa           core.PolicyAuthority
+	dbMap        *gorp.DbMap
+	certs        chan core.Certificate
+	issuedReport report
+}
+
+func newChecker(dbMap *gorp.DbMap) certChecker {
+	c := certChecker{
+		pa:    policy.NewPolicyAuthorityImpl(),
+		dbMap: dbMap,
+		certs: make(chan core.Certificate, batchSize),
+	}
+	c.issuedReport.Entries = make(map[string]reportEntry)
+	return c
+}
+
+func (c *certChecker) getCerts() error {
+	tx, err := c.dbMap.Begin()
+	if err != nil {
+		return err
+	}
+
+	cp := time.Now().Add(-checkPeriod)
+
+	var count int
+	err = c.dbMap.SelectOne(
+		&count,
+		"SELECT count(*) FROM certificates WHERE issued >= :issued",
+		map[string]interface{}{"issued": cp},
+	)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	// Retrieve certs in batches of 100000 (the size of the certificate channel)
+	// so that we don't eat unnecessary amounts of memory
+	for offset := 0; offset < count; {
+		var certs []core.Certificate
+		_, err = c.dbMap.Select(
+			&certs,
+			"SELECT * FROM certificates WHERE issued >= :issued ORDER BY issued ASC LIMIT :limit OFFSET :offset",
```

Comment:
Note: This query will need to be reworked to use indexes more efficiently. Given that that's going to be a bit invasive, and this is an audit-critical PR, I'm willing to wait on that. Please file an issue and link to it in a TODO.

We have a couple options: We could just add an index on issued, but that's wasteful for a query we'll run four times a year. Alternately: we run the one expensive, unindexed query to find the first certificate issued during the time span, and then do each query as `serial > :lastSerial order by serial ASC`, where `:lastSerial` is the highest serial in the last batch. Downside: This doubles down our dependency on a monotonically increasing serial prefix. Another alternate: We could just accept that this query, run 4 times a year, will be slow.